servicetalk-grpc-health/src/main/java/io/servicetalk/grpc/health/DefaultHealthService.java

```diff
@@ -0,0 +1,210 @@
+import io.servicetalk.health.v1.HealthCheckRequest;
+import io.servicetalk.health.v1.HealthCheckResponse;
+import io.servicetalk.health.v1.HealthCheckResponse.ServingStatus;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Predicate;
+
+import static io.servicetalk.concurrent.api.Processors.newPublisherProcessorDropHeadOnOverflow;
+import static io.servicetalk.concurrent.api.SourceAdapters.fromSource;
+import static io.servicetalk.grpc.api.GrpcStatusCode.FAILED_PRECONDITION;
+import static io.servicetalk.grpc.api.GrpcStatusCode.NOT_FOUND;
+import static io.servicetalk.health.v1.HealthCheckResponse.ServingStatus.NOT_SERVING;
+import static io.servicetalk.health.v1.HealthCheckResponse.ServingStatus.SERVICE_UNKNOWN;
+import static io.servicetalk.health.v1.HealthCheckResponse.ServingStatus.SERVING;
+import static io.servicetalk.health.v1.HealthCheckResponse.newBuilder;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Implementation of {@link Health.HealthService} which targets
+ * <a href="https://github.com/grpc/grpc/blob/master/doc/health-checking.md">gRPC health checking</a> that provides
+ * accessors to set/clear status for arbitrary services.
+ */
+public final class DefaultHealthService implements Health.HealthService {
+    /**
+     * The name of the service corresponding to
+     * the <a href="https://github.com/grpc/grpc/blob/master/doc/health-checking.md">overall health status</a>.
+     */
+    public static final String OVERALL_SERVICE_NAME = "";
+    private final Map<String, HealthValue> serviceToStatusMap = new ConcurrentHashMap<>();
+    private final Predicate<String> watchAllowed;
+    private final Lock lock = new ReentrantLock();
+    private boolean terminated;
+
+    /**
+     * Create a new instance. Service {@link #OVERALL_SERVICE_NAME} state is set to {@link ServingStatus#SERVING}.
+     */
+    public DefaultHealthService() {
+        this(service -> true);
+    }
+
+    /**
+     * Create a new instance. Service {@link #OVERALL_SERVICE_NAME} state is set to {@link ServingStatus#SERVING}.
+     * @param watchAllowed {@link Predicate} that determines if a {@link #watch(GrpcServiceContext, HealthCheckRequest)}
+     * request that doesn't match an existing service will succeed or fail with
+     * {@link GrpcStatusCode#FAILED_PRECONDITION}. This can be used to bound memory by restricting watches to expected
+     * service names.
+     */
+    public DefaultHealthService(Predicate<String> watchAllowed) {
+        this.watchAllowed = requireNonNull(watchAllowed);
+        serviceToStatusMap.put(OVERALL_SERVICE_NAME, new HealthValue(SERVING));
+    }
+
+    @Override
+    public Single<HealthCheckResponse> check(final GrpcServiceContext ctx, final HealthCheckRequest request) {
+        HealthValue health = serviceToStatusMap.get(request.getService());
```

Comment:
> Currently, users may see SERVING after it actually was terminated.

This shouldn't happen as terminated will send `NOT_SERVING` to all services and then clear the map. There is some concurrency but the map is `ConcurrentHashMap` so the map modification should be eventually visible.

Reply:
The problem is if users use only `check` without `watch`. After `terminated` is invoked, concurrent `check` may return `SERVING`. The follow-up `check` will return `grpc-status=NOT_FOUND`. This will be hard to interpret on the caller side. We should return `NOT_SERVING` after `terminated` for all services.

Reply:
grpc-java does not clear the map to keep returning `NOT_SERVING` after termination

Reply:
ah I see, that is a different topic, and good catch. I can not clear the map and then folks will get `NOT_SERVING` in a persistent fashion which is consistent with grpc-java.